plugins/hls-change-type-signature-plugin/src/Ide/Plugin/ChangeTypeSignature.hs

```diff
@@ -0,0 +1,247 @@
+      nfp <- getNormalizedFilePath plId (TextDocumentIdentifier uri)
+      decls <- getDecls ideState nfp
+      let actions = mapMaybe (generateAction uri decls) diags
+      pure $ List actions
+
+getDecls :: MonadIO m => IdeState -> NormalizedFilePath -> ExceptT String m [LHsDecl GhcPs]
+getDecls state = handleMaybeM "Error: Could not get Parsed Module"
+    . liftIO
+    . fmap (fmap (hsmodDecls . unLoc . pm_parsed_source))
+    . runAction "changeSignature.GetParsedModule" state
+    . use GetParsedModule
+
+-- | Text representing a Declaration's Name
+type DeclName = Text
+-- | The signature provided by GHC Error Message (Expected type)
+type ExpectedSig = Text
+-- | The signature provided by GHC Error Message (Actual type)
+type ActualSig = Text
+-- | The signature defined by the user
+type DefinedSig = Text
+
+-- | DataType that encodes the necessary information for changing a type signature
+data ChangeSignature = ChangeSignature {
+                         -- | The expected type based on Signature
+                         expectedType :: ExpectedSig
+                                       ,
+                         -- | the Actual Type based on definition
+                         actualType   :: ActualSig
+                                       ,
+                         -- | the declaration name to be updated
+                         declName     :: DeclName
+                                       ,
+                         -- | the location of the declaration signature
+                         declSrcSpan  :: RealSrcSpan
+                                       ,
+                         -- | the diagnostic to solve
+                         diagnostic   :: Diagnostic
+                                       }
+
+-- | Constraint needed to trackdown OccNames in signatures
+type SigName = (HasOccName (IdP GhcPs))
+
+
+-- | Create a CodeAction from a Diagnostic
+generateAction :: SigName => Uri -> [LHsDecl GhcPs] -> Diagnostic -> Maybe (Command |? CodeAction)
+generateAction uri decls diag = changeSigToCodeAction uri <$> diagnosticToChangeSig decls diag
+
+-- | Convert a diagnostic into a ChangeSignature and add the proper SrcSpan
+diagnosticToChangeSig :: SigName => [LHsDecl GhcPs] -> Diagnostic -> Maybe ChangeSignature
+diagnosticToChangeSig decls diagnostic = do
+    -- regex match on the GHC Error Message
+    (expectedType, actualType', declName) <- matchingDiagnostic diagnostic
+    let actualType = tidyTypeSignature actualType'
+    -- Find the definition and it's location
+    (declSrcSpan, ghcSig) <- findSigLocOfStringDecl decls (T.unpack declName)
+    -- Make sure the given "Actual Type" is a full signature
+    isValidMessage expectedType ghcSig
+    pure $ ChangeSignature{..}
+
+-- | Does the GHC Error Message give us a Signature we can use?
+-- We only want to change signatures when the "expected signature" given by GHC
+-- matches the one given by the user
+isValidMessage :: DefinedSig -> ExpectedSig -> Maybe ()
```

Comment:
I think slightly more idiomatic to return a bool and use `guard` at the call site.